### ISSUE TITLE
Add abort (or similar) to HttpClientRequest

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#22265]
HttpRequest (in dart:html) has an abort() method

There is a request to add a feature to the http package to support timeout.

This can be done with dart:html, but not dart:io

**Comment by mit-mit** [comment]
@sgjesse is anyone actually working on this? If not, should we close it as not planned?

**Comment by sgjesse** [comment]
At the moment we don't have a good solution to this. Using `Future.timeout` on in the various steps should work. That will not cancel the underlying operation, but that will fail in the background.

**Comment by davenotik** [comment]
Wow, still no timeout support. We'll continue using our custom `httpRead()` method then. Here it is:

https://gist.github.com/davenotik/169cc3a00d428516b257

It uses a Completer to build and return a Future that we can complete (time out) later.

**Comment by frank-rollpin** [comment]
Just wondering if there are any plans to address it, or should we forget about it and use workarounds?

**Comment by butlermatt** [comment]
Now that Timeout has been added to Socket.connect can a similar timeout be added to HttpClientRequest. I have an issue where I make a connection to a remote host, the connection is established but then the remote host gets "busy" for 10 minutes and blocks the event loop entirely. After that period the remote server eventually closes the connection and I get an error "Remote host closed connection before full headers were received" or something like that.

If I try adding a `.timeout` to the future from the HttpClientRequest.close() the timeout exception (set for 10 seconds) still waits for the ten minutes for the connection to fail because once established it is blocking.

**Comment by kevmoo** [comment]
Likely blocked on some refactoring we're doing with `dart:io` in Dart 2.

But certainly possible.

**Comment by sir-boformer** [comment]
Are there any workarounds that can be used to cancel a request?

I'm downloading large files from a server and sometimes users may want to cancel the request so that the download or upload is actually stopped.

**Comment by qwilbird** [comment]
+1

We are facing the same issue and we cannot release to prodution because of this. Any chance this will get fixed?

**Comment by kevmoo** [comment]
@zanderso @a-siva @mit-mit – thoughts on this?

**Comment by zanderso** [comment]
The new static field `HttpClient.connectionTimeout` (https://api.dartlang.org/stable/2.0.0/dart-io/HttpClient/connectionTimeout.html) may help with some component of this, but I'm not sure it will cover everything described here.

**Comment by sir-boformer** [comment]
This is still a critical shortcoming for any app that supports file downloads or other long-running http requests. A connection timeout is not really a solution.

**Comment by kevmoo** [comment]
@a-siva @mit-mit @zanderso – I'd love to push on this some more. We really should get on this.

**Comment by mit-mit** [comment]
cc @sortie who's looking into `:io`

**Comment by jaumard** [comment]
Any news on this ?? I'm doing a MJPEG reader, if I can't abort the request does this mean once it start it will actually never stop ??

**Comment by kevmoo** [comment]
Or @jonasfj ?

**Comment by jonasfj** [comment]
I haven't tried it, but if you're streaming down data from a response, you can probably stop reading the stream and call [`/HttpClientResponse.detachSocket`](https://api.dartlang.org/stable/2.3.2/dart-io/HttpClientResponse/detachSocket.html) which gives you a socket you can [`destroy`](https://api.dartlang.org/stable/2.3.2/dart-io/Socket/destroy.html).

But yes, there seems to be something missing here. Both on request and response.

**Comment by mit-mit** [comment]
cc @sortie

**Comment by mnordine** [comment]
Any update on this?

**Comment by mit-mit** [comment]
cc @zichangg

**Comment by PavelPZ** [comment]
I can see three infos regarding timeouts in this thread, how they are related? 
- @butlermatt : Now that Timeout has been added to Socket.connect...
- @zanderso : new static field HttpClient.connectionTimeout may help with some component...
- @jonasfj : call HttpClientResponse.detachSocket which gives you a socket you can destroy...

```package:http``` with timeout is blocking feature for me. As I understand, problem is with dart:io (dart:html has solution for it).

Some example with the best solution so far would be very helpful.

**Comment by a-siva** [comment]
zichang@ should be able to look at this once he is back from vacation.

**Comment by j0nscalet** [comment]
Seems like [CancellableOperation](https://pub.dev/documentation/async/latest/async/CancelableOperation-class.html) could be helpful here? A potential solution route?

**Comment by PavelPZ** [comment]
@ZichangG, I am still waiting for explanation if https://github.com/dart-lang/sdk/issues/22265#issuecomment-571626796

**Comment by kevmoo** [comment]
@j0nscalet – sadly not. In many cases you want the network connection to terminate. The CancellableOperation won't do that – although it will give you back control flow

**Comment by j0nscalet** [comment]
@kevmoo Ah right, makes sense. Thanks.

**Comment by ZichangG** [comment]
[connectionTimeout](https://api.dart.dev/stable/2.7.1/dart-io/HttpClient/connectionTimeout.html) is probably the answer. Here is an example.

```dart
  HttpClient client = new HttpClient();
  // Set up a time limit.
  client.connectionTimeout = Duration(microseconds: 1);

  client.getUrl(Uri.parse("https://www.google.com/"))
   .then((HttpClientRequest request) {
     // Optionally set up headers...
     // Optionally write to the request object...
     // Then call close.
     return request.close();
   })
   .then((HttpClientResponse response) {
     // response may not come back, so detachSocket() is not reliable.
   });
```

For other methods like `Socket.connect` and `detachSocket` are not good solutions for this problem. `Socket.destroy` will close the connection but it will not be exposed to user unless `detachSocket` is called. Since everything here is connected with `Future.then` mechanism. The only place to call `detachSocket` to abort the connection is when `response` is received. (It doesn't make any sense to cancel after request.close().) In this case, `HttpClientResponse` may not come back and `detachSocket` will not be called.

Using `Future.timeout` or `CancelableOperation ` is probably not closing the underlying connection.

For providing a timeout handler, perhaps a `try catch` block can be used to catch timeout exception.
```dart
try {
  // Http connections.
} catch (e) {
  // Check exception type and provide handler.
}
```

**Comment by januwA** [comment]
There is no good solution yet

**Comment by natebosch** [comment]
#388 added a way to call `detachSocket` as long as you are using an `IOClient` and using a streamed response.

If we can find a way to accomplish something similar on the web we might be able to expose some new API.

**Comment by lirantzairi** [comment]
I'm also looking for a way to do this. I have a project where I need to upload images to the server, which of course can take time depending on the size of the image. I want provide the user a way to cancel the operation in case it's taking too long without the upload request continuing in the background. Can anybody help with a workaround? Thanks.

**Comment by zahidaz** [comment]
it's almost five years and still looking for a standard solution, :(

**Comment by kevmoo** [comment]
Five years later! Never say never!

https://github.com/dart-lang/sdk/commit/4b96f20a79a6eca074dc17b8b84448f79d4acfc8

**Comment by vsmenon** [comment]
Is the plan to reland the above CL?

**Comment by zichangg** [comment]
Reland was ready! Because Lasse went on vacation. He will be back next week and I'll get it land next week.

**Comment by vsmenon** [comment]
Thanks @zichangg - marking this as part of Sept release for now.

**Comment by mit-mit** [comment]
@zichangg looks like this was relanded?

**Comment by zichangg** [comment]
Yes. It has fixed the broken test and is ready to go.

**Comment by mit-mit** [comment]
Related issue for exposing this in `package:http`: https://github.com/dart-lang/sdk/issues/22265
